File: client/lib/src/envelope/envelope_codec.dart
import 'dart:typed_data';

import '../crypto/canonical_cbor.dart';
import 'envelope_rejection.dart';
import 'shared_envelope.dart';

/// Encodes and decodes [SharedEnvelope] to/from canonical CBOR.
class EnvelopeCodec {
  final CanonicalCborMapper _cbor = CanonicalCborMapper();

  /// Encodes [envelope] to canonical CBOR, including [signature] if present.
  Uint8List encode(SharedEnvelope envelope) {
    final map = _toMap(envelope);
    if (envelope.signature != null) {
      map['signature'] = envelope.signature!;
    }
    return _cbor.writeBytes(map);
  }

  /// Decodes [bytes] into a [SharedEnvelope].
  ///
  /// Throws [MalformedCbor] if the bytes cannot be parsed or are non-canonical.
  SharedEnvelope decode(Uint8List bytes) {
    Map<String, dynamic> map;
    try {
      map = _cbor.readValue(bytes);
    } catch (e) {
      throw MalformedCbor('Failed to parse envelope CBOR', e);
    }
    return _fromMap(map);
  }

  /// Returns the bytes that are covered by the issuer signature (no signature field).
  Uint8List signedBytesOf(SharedEnvelope envelope) {
    return _cbor.writeBytes(_toMap(envelope));
  }

  Map<String, Object?> _toMap(SharedEnvelope env) {
    final map = <String, Object?>{};
    map['version'] = env.version;
    map['entry_id'] = env.entryId;
    map['issuer_id'] = env.issuerId;
    map['issuer_label'] = env.issuerLabel;
    map['issuer_signing_key_id'] = env.issuerSigningKeyId;
    map['created_at'] = env.createdAt;
    map['description'] = env.description;
    map['ciphertext_alg'] = env.ciphertextAlg;
    map['ciphertext_nonce'] = env.ciphertextNonce;
    map['ciphertext'] = env.ciphertext;
    map['ciphertext_hash'] = env.ciphertextHash;

    final wrappings = env.recipientWrappings.map((rw) {
      return <String, Object?>{
        'verifier_id': rw.verifierId,
        'verifier_key_id': rw.verifierKeyId,
        'wrapped_data_key': rw.wrappedDataKey,
      };
    }).toList();
    map['recipient_wrappings'] = wrappings;
    return map;
  }

  SharedEnvelope _fromMap(Map<String, dynamic> map) {
    final version = (map['version'] as int);
    final entryId = map['entry_id'] as Uint8List;
    final issuerId = map['issuer_id'] as Uint8List;
    final issuerLabel = map['issuer_label'] as String;
    final issuerSigningKeyId = map['issuer_signing_key_id'] as Uint8List;
    final createdAt = map['created_at'] as int;
    final description = map['description'] as String;
    final ciphertextAlg = map['ciphertext_alg'] as String;
    final ciphertextNonce = map['ciphertext_nonce'] as Uint8List;
    final ciphertext = map['ciphertext'] as Uint8List;
    final ciphertextHash = map['ciphertext_hash'] as Uint8List;
    final signature = map['signature'] as Uint8List?;

    final rawWrappings = map['recipient_wrappings'] as List<dynamic>;
    final wrappings = rawWrappings.map((rw) {
      final rwMap = rw as Map<String, dynamic>;
      return RecipientWrapping(
        verifierId: rwMap['verifier_id'] as Uint8List,
        verifierKeyId: rwMap['verifier_key_id'] as Uint8List,
        wrappedDataKey: rwMap['wrapped_data_key'] as Uint8List,
      );
    }).toList();

    return SharedEnvelope(
      version: version,
      entryId: entryId,
      issuerId: issuerId,
      issuerLabel: issuerLabel,
      issuerSigningKeyId: issuerSigningKeyId,
      createdAt: createdAt,
      description: description,
      ciphertextAlg: ciphertextAlg,
      ciphertextNonce: ciphertextNonce,
      ciphertext: ciphertext,
      ciphertextHash: ciphertextHash,
      recipientWrappings: wrappings,
      signature: signature,
    );
  }
}

File: src/main/java/com/wilhelmsen/cbslink/plugin/datawallet/envelope/EnvelopeCodec.java
package com.wilhelmsen.cbslink.plugin.datawallet.envelope;

import com.wilhelmsen.cbslink.plugin.datawallet.crypto.CanonicalCborMapper;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public final class EnvelopeCodec {

    private final CanonicalCborMapper cbor = new CanonicalCborMapper();

    public byte[] encode(SharedEnvelope envelope) {
        Map<String, Object> map = toMap(envelope);
        if (envelope.signature() != null) {
            map.put("signature", envelope.signature());
        }
        return cbor.writeBytes(map);
    }

    @SuppressWarnings("unchecked")
    public SharedEnvelope decode(byte[] bytes) {
        Map<String, Object> map;
        try {
            map = cbor.readValue(bytes, Map.class);
        } catch (Exception e) {
            throw new EnvelopeRejection.MalformedCbor("Failed to parse envelope CBOR", e);
        }
        return fromMap(map);
    }

    public byte[] signedBytesOf(SharedEnvelope envelope) {
        Map<String, Object> map = toMap(envelope);
        return cbor.writeBytes(map);
    }

    private Map<String, Object> toMap(SharedEnvelope env) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("version", env.version());
        map.put("entry_id", uuidToBytes(env.entryId()));
        map.put("issuer_id", uuidToBytes(env.issuerId()));
        map.put("issuer_label", env.issuerLabel());
        map.put("issuer_signing_key_id", env.issuerSigningKeyId());
        map.put("created_at", env.createdAt());
        map.put("description", env.description());
        map.put("ciphertext_alg", env.ciphertextAlg());
        map.put("ciphertext_nonce", env.ciphertextNonce());
        map.put("ciphertext", env.ciphertext());
        map.put("ciphertext_hash", env.ciphertextHash());

        List<Map<String, Object>> wrappings = new ArrayList<>();
        for (RecipientWrapping rw : env.recipientWrappings()) {
            Map<String, Object> rwMap = new LinkedHashMap<>();
            rwMap.put("verifier_id", uuidToBytes(rw.verifierId()));
            rwMap.put("verifier_key_id", rw.verifierKeyId());
            rwMap.put("wrapped_data_key", rw.wrappedDataKey());
            wrappings.add(rwMap);
        }
        map.put("recipient_wrappings", wrappings);
        return map;
    }

    @SuppressWarnings("unchecked")
    private SharedEnvelope fromMap(Map<String, Object> map) {
        int version = ((Number) map.get("version")).intValue();
        UUID entryId = bytesToUuid((byte[]) map.get("entry_id"));
        UUID issuerId = bytesToUuid((byte[]) map.get("issuer_id"));
        String issuerLabel = (String) map.get("issuer_label");
        byte[] issuerSigningKeyId = (byte[]) map.get("issuer_signing_key_id");
        long createdAt = ((Number) map.get("created_at")).longValue();
        String description = (String) map.get("description");
        String ciphertextAlg = (String) map.get("ciphertext_alg");
        byte[] ciphertextNonce = (byte[]) map.get("ciphertext_nonce");
        byte[] ciphertext = (byte[]) map.get("ciphertext");
        byte[] ciphertextHash = (byte[]) map.get("ciphertext_hash");
        byte[] signature = (byte[]) map.get("signature");

        List<Map<String, Object>> rawWrappings = (List<Map<String, Object>>) map.get("recipient_wrappings");
        List<RecipientWrapping> wrappings = new ArrayList<>();
        for (Map<String, Object> rw : rawWrappings) {
            wrappings.add(new RecipientWrapping(
                    bytesToUuid((byte[]) rw.get("verifier_id")),
                    (byte[]) rw.get("verifier_key_id"),
                    (byte[]) rw.get("wrapped_data_key")
            ));
        }

        return new SharedEnvelope(
                version, entryId, issuerId, issuerLabel, issuerSigningKeyId,
                createdAt, description, ciphertextAlg, ciphertextNonce,
                ciphertext, ciphertextHash, wrappings, signature
        );
    }

    static byte[] uuidToBytes(UUID uuid) {
        byte[] bytes = new byte[16];
        ByteBuffer buf = ByteBuffer.wrap(bytes);
        buf.putLong(uuid.getMostSignificantBits());
        buf.putLong(uuid.getLeastSignificantBits());
        return bytes;
    }

    static UUID bytesToUuid(byte[] bytes) {
        ByteBuffer buf = ByteBuffer.wrap(bytes);
        return new UUID(buf.getLong(), buf.getLong());
    }
}
